File: 2022/go/day7/day7.go
package main

import (
	"fmt"
	"os"
	"strings"
)

func main() {
	input, _ := os.ReadFile("../../input/day7.txt")
	root := parseTerminal(string(input))
	var part1 int
	root.forEach(func(n *node) {
		if n.fileSize == 0 && n.size() <= 100000 {
			part1 += n.size()
		}
	})
	fmt.Printf("part1: %v\n", part1)

	needed := 30000000 - (70000000 - root.size())

	smallest := 70000000
	root.forEach(func(n *node) {
		if n.size() >= needed && n.size() < smallest {
			smallest = n.size()
		}
	})
	fmt.Printf("part2: %v\n", smallest)
}

type node struct {
	fileSize int
	parent   *node
	nodes    map[string]*node
}

func parseTerminal(input string) *node {
	root := newNode(0, nil)
	cwd := root
	lines := strings.Split(input, "\n")
	for _, line := range lines {
		cwd = cwd.apply(line)
	}
	return root
}

func (n *node) apply(line string) *node {
	name, size := "", 0
	if strings.HasPrefix(line, "$ cd ") {
		return n.cd(line[5:])
	} else if _, err := fmt.Sscanf(line, "%d %s", &size, &name); err == nil {
		return n.add(name, size)
	}
	return n
}

func (n *node) cd(name string) *node {
	if name == "/" && n.parent == nil {
		return n
	} else if name == "/" {
		return n.cd("/")
	} else if name == ".." {
		return n.parent
	}
	if _, ok := n.nodes[name]; !ok {
		n.nodes[name] = newNode(0, n)
	}
	return n.nodes[name]
}

func (n *node) add(name string, size int) *node {
	n.nodes[name] = newNode(size, n)
	return n
}

func (n *node) size() int {
	if n.fileSize > 0 {
		return n.fileSize
	}
	var sum int
	for _, c := range n.nodes {
		sum += c.size()
	}
	return sum
}

func newNode(size int, parent *node) *node {
	return &node{fileSize: size, parent: parent, nodes: make(map[string]*node)}
}

func (n *node) forEach(f func(*node)) {
	for _, c := range n.nodes {
		f(c)
		c.forEach(f)
	}
}

File: 2022/java/Day7.java
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Map;
import java.util.TreeMap;
import java.util.regex.Pattern;
import java.util.stream.Stream;

class Day7 {
    public static void main(String ... args) throws Exception {
        var root = parseTerminal(Files.readString(Path.of("../input/day7.txt")));

        var part1 = root.dirs()
            .filter(n -> n.size() <= 100000)
            .mapToInt(n -> n.size()).sum();
        System.out.printf("part 1: %s\n", part1);

        var needed = 30000000 - (70000000 - root.size());
        var part2 = root.dirs()
            .filter(n -> n.size() >= needed)
            .mapToInt(n -> n.size()).sorted().findFirst().getAsInt();
        System.out.printf("part 2: %s\n", part2);
    }

    static Node parseTerminal(String input) {
        Node root = new Node(0, null, new TreeMap<>());
        input.lines().reduce(root, (n, l) -> n.apply(l), (a,b) -> {return null;});
        return root;
    }
}

record Node(int size, Node parent, Map<String,Node> nodes) {
    static Pattern filePattern = Pattern.compile("(\\d+) ([\\w.]+)");

    Node apply(String line) {
        if (line.startsWith("$ cd ")) {
            return cd(line.substring(5));
        }
        var fileMatcher = filePattern.matcher(line);
        if (fileMatcher.matches()) {
            return add(fileMatcher.group(2), fileMatcher.group(1));
        }
        return this;
    }

    Node cd(String name) {
        return switch (name) {
            case "/" -> parent == null ? this : parent.cd("/");
            case ".." -> parent;
            default -> nodes.computeIfAbsent(name, k -> new Node(0, this, new TreeMap<>()));
        };
    }

    Node add(String name, String size) {
        nodes.put(name, new Node(Integer.parseInt(size), this, new TreeMap<>()));
        return this;
    }

    Stream<Node> dirs() {
        return Stream.concat(Stream.of(this), nodes.values().stream().flatMap(n -> n.dirs())
            .filter(n -> n.size == 0));
    }

    public int size() {
        return size > 0 ? size : nodes.values().stream().mapToInt(n -> n.size()).sum();
    }
};